### ISSUE TITLE
Simple question about Netmiko and multithreading

### ISSUE THREAD

**Issue by ahibbard-cornelisnetworks** [ktbyers/netmiko#2999]
I have a simple question about Netmiko whether it is thread safe. I know that Paramiko is the base of Netmiko and not thread safe, but I have also seen multiple examples online of people multithreading Netmiko. So, is it safe to multithread Netmiko?

**Comment by ktbyers** [comment]
It is "probably" fine as long as each thread is managing a separate SSH connection.

I would (probably) not try to share a single Netmiko connection among multiple threads. 

The first case is what almost everyone is doing.

**Comment by ahibbard-cornelisnetworks** [comment]
Hey, thanks for the quick response. Another quick question, I tried to use the link in the Readme to join the Netmiko Slack channel, but got a message that the link is no longer active. Any chance I could get an invite?

**Comment by schautwb** [comment]
@ahibbard-cornelisnetworks Here's a link to join Slack: [https://join.slack.com/t/pynet/shared_invite/zt-1in1qx0nl-XShKPynlo8dnI6dzOkZY7Q](url)

**Comment by HBLiB** [comment]
When i go through a list of 500 devices, and even if I use 10 threads ( above 15 threads it definitely happens).
That i run in to OS limitations, too many open files, even though the socket times out ( netstat -tulpen) and I do not see the process (you can actually see the ssh process towards the host) after its done with a particular device. The PIPE remains open (visible with lsof ) . I can make it a bit workable with increasing this limit with lsof -n 9000. But still can't go over the 10 threads a.k.a 10 devices at a time limit when going through a list of 100+ devices.

So my experience is, when using threads with < 60-80 devices, i go with 20 threads.
but anything above 80 devices, I do not take a risk and run it with 10 threads.

But when not using threading, the script goes through 600 devices in around 50-55 minutes, without issues.

Not sure if I;m expecting too much and/or anybody run in to this issue too

**Comment by ktbyers** [comment]
@HBLiB 

```
That i run in to OS limitations, too many open files, even though the socket times out
```

1. Are you making sure you disconnect/use a context manager for the connections.
2. Have you used `ulimit` to expand the allowed open files. Note, TCP connections will be counted as open files.
3. I would use `netstat -an`. On my Linux system `netstat -tulpen` does not give me all of the connections.


```
I do not see the process
```

Not sure I followed this, you are using threads so you will only have one process (i.e. the single Python process) that spawns all of the threads.

```
Are you using AAA (particularly authorization) on your network devices?
```

You should be able to use a lot more than 10 threads that seems artificially low to me.

The most common cause of issues similar to what you describe is AAA not being able to handle the concurrency load. This causes the devices to slow way down (i.e. AAA is the bottleneck and can't handle the load presented by the threads). This happens a lot (though may or may not be relevant to you).

Probably need a bit more info, but on the face of it definitely sounds like something that should be able to be corrected (assuming that it is not AAA).

**Comment by HBLiB** [comment]
@ktbyers 

> * Are you making sure you disconnect/use a context manager for the connections.

Yes, when using ConnectHandler I disconnect everyhost. And SSHDetect I cant't as far as I could find at least.

> * Have you used `ulimit` to expand the allowed open files. Note, TCP connections will be counted as open files.

That is indeed how I was able to make it perform on longer device lists

> * I would use `netstat -an`. On my Linux system `netstat -tulpen` does not give me all of the connections.

As i was particulalry interested in tcp sockets and in their state (ESTABLISHED/TIMEOUT), I used -tulpen. maybe I could check the unix sockets too to debug more. But seems pretty clear its the open files/socket/entries in the lsof list.

>Not sure I followed this, you are using threads so you will only have one process (i.e. the single Python process) that spawns all of the threads.

When running netmikko ConnectHandler or SSHDetect i can see the ssh command in the output of ps aux on my ubuntu machine.


> The most common cause of issues similar to what you describe is AAA not being able to handle the concurrency load. This causes the devices to slow way down (i.e. AAA is the bottleneck and can't handle the load presented by the threads). This happens a lot (though may or may not be relevant to you).

Im sure its not AAA or connection issues. The exception i receive is the "too many files open" error.
But i have also sniffed the traffic, when it start too fail with too many file open error, it doesnt even try to open the socket due to the limit. Hence the ulimit "fix".
And as i mentioned it also only happes when the device list is very long.
If i run 50 devices in 50 threads its "fine".
But when i run 537 devices with 20 threads. I receive the 'too many' files open error after maybe 300 devices.

**Comment by ktbyers** [comment]
Okay, I will have to look into SSHDetect and whether there is a way to gracefully disconnect that (but realistically that should be a one-off process for discovering device_type which should then be saved in some way).

Yeah, netstat -tulpen did not show me 'established' connections on my Linux system, but if you have another way of getting the connections that are still open...that is what matters. 

```
When running netmikko ConnectHandler or SSHDetect i can see the ssh command in the output of ps aux on my ubuntu machine.
```

Are you doing SSH proxy via an SSH config file or something else atypical (as all of the direct SSH connections via Netmiko/Paramiko are not visible at the system level i.e. neither Netmiko nor Paramiko uses the system SSH except in special cases namely SSH proxy)?

You are probably need to show your code (as you shouldn't be hitting the max files open issues at only 500 devices and certainly shouldn't hit it, if you increase you open files to 9000).

 In other words, even if you kept the connections open and never closed them, you shouldn't hit the 1024 threshold. Of course, I am making assumptions about what you are doing.

**Comment by HBLiB** [comment]
@ktbyers 
Sorry, I forgot to mention it is indeed through a proxy. I guess that could be the issue causing this.
And the reason im not seeing this with Paramiko is that I actually make aconnection to the jumphost per thread and from there go through the list. So you only have x constant ssh connections in relation to the amount of threads with the jumphost, rather than python initiating a new ssh session for everyhost in turn and causing a butload of lsof entries. (Just assuming for now.)

But I would prefer to have it through a proxy as you can "touch" different network/client environments and their jumphost not having proper libraries or not being up to date can prevent me from using "my" scripts. 

I have uploaded my code to [github](https://github.com/HBLiB/netmikoMP).
Really appreciate your time!

**Comment by HBLiB** [comment]
@ktbyers 
Not sure if you had time to look into the jumphost and separate connection per host thing.
In the meantime, I adjusted my socket code a bit and uploaded it to [github](https://github.com/HBLiB/netmikoMP/blob/main/SocketNetmiko)


I will try to fork it in Netmiko as an extra option, but can't promise anything as I am mainly a network engineer not a coder.. hehe
